designs/use-selectors-at-cache.md

```diff
@@ -0,0 +1,125 @@
+Filter cache ListWatch using selectors
+===================
+## Motivation
+
+Controller-Runtime controllers use a cache to subscribe to events from
+Kubernetes objects and to read those objects more efficiently by avoiding
+to call out to the API. This cache is backed by Kubernetes informers.
+
+The only way to filter this cache is by namespace and resource type.
+In cases where a controller is only interested in a small subset of objects
+(for example all pods on a node), this might end up not being efficient enough.
+
+Requests to a client backed by a filtered cache for objects that do not match
+the filter will never return anything, so we need to make sure that we properly
+warn users to only use this when they are sure they know what they are doing.
+
+This proposal sidesteps the issue of "How to we plug this into the cache-backed
+client so that users get feedback when they request something that is
+not matching the caches filter" by only implementing the filter logic in the
+cache package. This allows advanced users to combine a filtered cache with the
+already existing `NewCacheFunc` option in the manager and cluster package,
+while simultaneously hiding it from newer users that might not be aware of the
+implications and the associated foot-shoot potential.
+
+The only alternative today to get a filtered cache with controller-runtime is
+to build it out-of tree. Because such a cache would mostly copy the existing
+cache and add just some options, this is not great for consumers.
+
+This proposal is related to the following issue [2]
+
+## Proposal
+
+Add a new selector code at `pkg/cache/internal/selector.go` with common structs
+and helpers
+```golang
+package internal
+
+...
+
+// SelectorsByObject associate a client.Object to a field/label selector
+type SelectorsByObject map[client.Object]Selector
+
+// SelectorsByObject associate a GroupResource to a field/label selector
+type SelectorsByGVK map[schema.GVK]Selector
+
+// Selector specify the label/field selector to fill in ListOptions
+type Selector struct {
+	Label labels.Selector
+	Field fields.Selector
+}
+
+// FillInListOpts fill in ListOptions LabelSelector and FieldSelector if needed
+func (s Selector) FillInListOpts(listOpts *metav1.ListOptions) {
+...
+}
+
+
+Add a type alias to `pkg/cache/cache.go` to internal
+
+```golang
+
+type SelectorsByObject internal.SelectorsByObject
+
+```
+
+Extend `cache.Options` as follows:
+
+```golang
+type Options struct {
+	Scheme            *runtime.Scheme
+	Mapper            meta.RESTMapper
+	Resync            *time.Duration
+	Namespace         string
+	SelectorsByObject SelectorsByObject
```

Comment:
It would be nice to generalize this to something like 
```suggestion
	Selectors SelectorSet
```
Where `SelectorSet` is any set of selectors that implements
```go
type SelectorSet interface {
	client.ListOption
	Select()
}
```
where `Select()` is just an identifier method so users don't pass in other `client.ListOption`s.

Reply:
Done

Reply:
Do you mean ?

```golang
type SelectorSet interface {
	Select(runtime.Object) client.ListOption
}
```

Reply:
We have apply options per GVK so it doe not make sense to implement ListOption at the map level.

Reply:
Sure, get rid of `client.ListOption` from that interface. Regardless, I may want to construct a GVK myself instead of using a typed object or `Unstructured`, so why not also expose `SelectorsByGVK`?

Reply:
As we have previously stated we are going to expose the map key as `client.Object`.